pandas/core/arrays/boolean.py

```diff
@@ -740,6 +742,162 @@ def boolean_arithmetic_method(self, other):
         return set_function_name(boolean_arithmetic_method, name, cls)
 
 
+def kleene_or(
+    left: Union[bool, np.nan, np.ndarray],
+    right: Union[bool, np.nan, np.ndarary],
+    left_mask: Optional[np.ndarary],
+    right_mask: Optional[np.ndarray],
+):
+    """
+    Boolean ``or`` using Kleene logic.
+
+    Values are NA where we have ``NA | NA`` or ``NA | False``.
+    ``NA | True`` is considered True.
+
+    Parameters
+    ----------
+    left, right : ndarray, NA, or bool
+        The values of the array.
+    left_mask, right_mask : ndarray, optional
+        The masks. When
+
+    Returns
+    -------
+    result, mask: ndarray[bool]
+        The result of the logical or, and the new mask.
+    """
+    # To reduce the number of cases, we ensure that `left` & `left_mask`
+    # always come from an array, not a scalar. This is safe, since because
+    # A | B == B | A
+    if left_mask is None:
+        return kleene_or(right, left, right_mask, left_mask)
+
+    assert left_mask is not None
+    right_is_scalar = right_mask is None
+
+    mask = left_mask
+
+    if right_mask is not None:
+        mask = mask | right_mask
+    else:
+        mask = mask.copy()
+
+    # handle scalars:
+    if right_is_scalar and np.isnan(right):  # TODO(pd.NA): change to NA
+        result = left.copy()
+        mask = left_mask.copy()
+        mask[~result] = True
+        return result, mask
+
+    # XXX: verify that this doesn't assume masked values are False!
+    result = left | right
+    mask[result] = False
```

Comment:
In the end, we want to know where either left or right was a True, right? 
But, that also depends on the mask, as the masked values should not be taken into account. 

So should this be something like `result = (left & ~left_mask) | (right & ~right_mask)`

(although there might be a more performant way with less steps)

Reply:
And then the final mask would be something like `(mask_left | mask_right) & ~result`

Reply:
But that gives many logical ops to perform one logical op ...

Reply:
Indeed, that's the downside.

For now, I think we just accept that cost. We attempt to ensure that masked values are always False, but that seems difficult to get correct, and becomes obsolete if / when we back these with Pyarrow arrays.

Reply:
> We attempt to ensure that masked values are always False, but that seems difficult to get correct

I don't think we should even try to attempt that